### PR TITLE
Feat: Server-side events for basenames claim frame

### PR DESCRIPTION
**Pull request by brendan-defi**
**What changed? Why?**
* created utility for logging server-side events
* implemented server-side event logging throughout basenames claim frame

**Notes to reviewers**
* I have been unable to get the events to work for the development config. Currently hardcoding to prod config.

**How has it been tested?**
locally

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: apps/web/pages/api/basenames/frame/03_getPriceAndConfirm.ts [hunk 1]
```diff
@@ -24,6 +27,22 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+  try {
+    const userAgent = req.headers['user-agent'] ?? 'No user agent';
+    const ip = req.headers['x-forwarded-for'] ?? req.socket.remoteAddress ?? 'No IP';
+    const deviceId = generateCustomUUID(userAgent, ip);
+    const eventName = 'selected_years';
+    const eventProperties = {
+      action: ActionType.click,
+      context: 'basenames_claim_frame',
+      componentType: ComponentType.button,
+    };
```
Comment (JFrankfurt) [inline apps/web/pages/api/basenames/frame/03_getPriceAndConfirm.ts:39]: it looks to me like this code could be deduplicated--what do you think?

Comment (brendan-defi) [reply apps/web/pages/api/basenames/frame/03_getPriceAndConfirm.ts:39]: makes sense. will update

File: apps/web/src/utils/logServerSideEvent.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  eventProperties: Omit<EventProperties, 'projectName'>,
+  supplementalEventData: SupplementalEventData = {},
+) {
+  const event = createEventData(eventName, deviceId, eventProperties, supplementalEventData);
+  const stringifiedEvent = JSON.stringify([convertKeys(event)]);
+  const uploadTime = new Date().getTime().toString();
+  const checksum = crypto
+    .createHash('md5')
+    .update(stringifiedEvent + uploadTime)
+    .digest('hex');
```
Comment (JFrankfurt) [inline apps/web/src/utils/logServerSideEvent.ts:75]: nit: I don't think checksumming is really necessary

File: apps/web/src/utils/logServerSideEvent.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  const fetchConfig = {
+    method: 'POST',
+    body: JSON.stringify(eventData),
+    mode: 'no-cors',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  };
+  fetch(postUrl, fetchConfig)
+    // .then((res) => console.log('Response Status:', res.status))
```
Comment (JFrankfurt) [inline apps/web/src/utils/logServerSideEvent.ts:91]:
```suggestion
```

File: apps/web/src/utils/logServerSideEvent.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+export function generateCustomUUID(userAgent: string, ipAddress: string) {
+  const data = `${userAgent}|${ipAddress}`;
+  const hash = crypto.createHash('md5').update(data).digest('hex');
+
+  return `${hash.substring(0, 8)}-${hash.substring(8, 4)}-${hash.substring(12, 4)}-${hash.substring(
+    16,
+    4,
+  )}-${hash.substring(20)}`;
```
Comment (JFrankfurt) [inline apps/web/src/utils/logServerSideEvent.ts:131]: nit: use an external lib

**Review by JFrankfurt — APPROVED** [review]

**Comment by cb-heimdall** [comment]
Review Error for JFrankfurt @ 2024-09-24 14:33:50 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

**Review by JFrankfurt — APPROVED** [review]

File: apps/web/src/utils/logServerSideEvent.ts (hunk 2 above)
Comment (brendan-defi) [reply apps/web/src/utils/logServerSideEvent.ts:75]: I'm not 100% sure it's necessary, but the CCA network call definitely includes this for client-side log requests, so I think safer to leave in if okay with you

File: apps/web/src/utils/logServerSideEvent.ts (hunk 3 above)
Comment (brendan-defi) [reply apps/web/src/utils/logServerSideEvent.ts:91]: removing

File: apps/web/src/utils/logServerSideEvent.ts (hunk 4 above)
Comment (brendan-defi) [reply apps/web/src/utils/logServerSideEvent.ts]: leveraged uuidV5 for this and checksum

**Pull request merged**